### DISCUSSION TITLE
Provide some keywords to set ODT metadata.

### DISCUSSION THREAD

**Discussion by hwiorn** [kjambunathan/org-mode-ox-odt#113] [Ideas]
Currently, exporter doesn't make general and custom properties of an ODT file. I can't set any general fields and add any custom fields inside an ODT document using ox-org only. I can define user custom fields in `#+ODT_MASTER_STYLES`, But I can't set them to metadata.

![image](https://user-images.githubusercontent.com/1394254/139083581-49bee0ba-8c06-4f41-b5dc-0d2911358d01.png)

![image](https://user-images.githubusercontent.com/1394254/139084059-107a1bdd-0ce4-44c9-a391-db1717385f8e.png)

I checked `ox-odt.el`, and I found it checks only [some keywords like title, author, description, etc](https://github.com/kjambunathan/org-mode-ox-odt/blob/master/lisp/ox-odt.el#L2874-L2946
).

I think that ox-odt needs to provide a custom property which can be overridden or `#+ODT_EXTRA_META` is a suitable solution. 

```org
#+property: created <2021-10-26 Tue>
#+property: modified <2021-10-27 Wed>
```
or
```org
#+ODT_GENERAL_META: modified <2021-10-26 Tue>
#+ODT_CUSTOM_META: Version 1.1.1
#+ODT_CUSTOM_META: Document-Id A-21-10-25-REQ-11
```
or
```org
#+ODT_EXTRA_META: <dc:date>2021-10-27T17:18:01.220502029</dc:date><meta:editing-duration>PT6M31S</meta:editing-duration><meta:editing-cycles>6</meta:editing-cycles><meta:generator>LibreOffice/7.2.2.2$Linux_X86_64 LibreOffice_project/20$Build-2</meta:generator>
```

**Comment by kjambunathan** [comment]
You can already export properties with `prop:t` option.  Look at an the Org snippet / screenshot example this issue: https://github.com/kjambunathan/org-mode-ox-odt/issues/95.

Can you give me more idea on your specific situation ....

1. What sort of user fields you want to put in meta.xml?  
2. Where do you intend to use those fields?  In document header / footer or inline within the document

I don't use properties or keywords much and at this point in time I have no clarity on what woud be the most flexible way to specify user fields ... keywords or properties?  If you have specific suggestions  I will consider them.

In principle I will take up your request for additional user fields .....

**Comment by hwiorn** [comment]
I'm trying to rewrite my model evaluation test results and technical support manuals to org file with ODT templates.
I found my issue is referencing variables(user field) in header and footer. 
I couldn't use the macro-replacement feature in `+CUSTOM_EXTRA_STYLE` so I used user-fields which can be controlled by custom properties in a document.

The one of my documents consists custom cover page which has no header and footer and other pages which is the general company document form which has speicific document information. 
I have to set `title:nil author:nil`. It means I need to define title alternative custom property for header.

#### The form of a header
```org
| IMG  |                                        <TITLE>                                                   | <Company |
| LOGO | Doc-Id | <DocID> | Version |  <DocVersion> | Created | <CreatedDate> | Modified | <ModifiedDate> | Name>    |
```

#### The form of a footer
```org
| <CurrentDate> |                   <PageNumber> / <PageCount>                          |  <CompanyName> <Department> |
```

**Reply by kjambunathan** [reply]
> Provide some keywords to set ODT metadata

I have added this feature.  See  https://github.com/kjambunathan/org-mode-ox-odt/issues/117 for more details.


> #### The form of a header
> 
> ```
> | IMG  |                                        <TITLE>                                                   | <Company |
> | LOGO | Doc-Id | <DocID> | Version |  <DocVersion> | Created | <CreatedDate> | Modified | <ModifiedDate> | Name>    |
> ```

If the document is in English, then the header will be too cramped.  If you are using korean or other east-asian languages this  may not be the case.

Here is a zip file for you to experiment with 
[docproperties.zip](https://github.com/kjambunathan/org-mode-ox-odt/files/7443154/docproperties.zip)

Note that I have used "transcluded table" (instead of list table).  Regarding alignment, I am not sure if you are talking about horizontal alignment or vertical alignment.  Horizontal alignment doesn't work right now in list tables or transcluded tables, because of https://github.com/kjambunathan/org-mode-ox-odt/issues/25.  I will remove handicap in next few days.  If you are talking about vertical alignment in tables, look at https://github.com/kjambunathan/org-mode-ox-odt/issues/110#issue-1031336353.  

I am not done with this discussion.  Experiment, and let me know what you think or want.

To get this,

![Screenshot from 2021-10-29 21-01-27](https://user-images.githubusercontent.com/19869864/139462588-2a4af678-d4ce-4b9b-b2c5-6f9b9a0fcca1.png)


do this,

```org
#+odt_preferred_output_format: pdf

#+begin_src emacs-lisp :exports results :results silent
(add-to-list 'org-odt-experimental-features 'transclude-sole-footnote-references-in-a-table)
#+end_src

#+ODT_DOCUMENT_PROPERTIES: DOC-TITLE DOC-COMPANY-NAME DOC-ID DOC-VERSION DOC-CREATED-DATE DOC-MODIFIED-DATE

#+DOC-TITLE:            Custom fields
#+DOC-COMPANY-NAME:     Org Mode ODT Exporter
#+DOC-ID:               ox-odt-8d6d991d-df0c-4aec-a548-ada78dd554ee
#+DOC-VERSION:          1.0
#+DOC-CREATED-DATE:     [2021-10-19 Tue]
#+DOC-MODIFIED-DATE:    <2021-10-29 Fri>

#+MACRO: DocTitle {{{keyword-to-documentproperty(DOC-TITLE)}}}
#+MACRO: CompanyName {{{keyword-to-documentproperty(DOC-COMPANY-NAME)}}}
#+MACRO: DocID {{{keyword-to-documentproperty(DOC-ID)}}}
#+MACRO: DocVersion {{{keyword-to-documentproperty(DOC-VERSION)}}}
#+MACRO: CreatedDate {{{keyword-to-documentproperty(DOC-CREATED-DATE)}}}
#+MACRO: ModifiedDate {{{keyword-to-documentproperty(DOC-MODIFIED-DATE)}}}

Nunc eleifend leo vitae magna.  Nunc eleifend leo vitae magna.
Curabitur lacinia pulvinar nibh.  Donec posuere augue in quam.
Aliquam posuere.  Nullam rutrum.  Aliquam posuere.

#+ATTR_ODT: :style "GriddedTable"
#+ATTR_ODT: :rel-width 100
#+ATTR_ODT: :widths "15, 8 , 8 , 8 , 8 , 8 , 8 , 8 , 8 , 8"
#+ATTR_ODT: :span "@1$1{3:1} @1$2{1:4} @1$6{3:1}"
| <l2>   |        <r1> | <1>              | <1>               | <1>                | <1>               |
| [fn:1] |      [fn:2] |                  |                   |                    | {{{CompanyName}}} |
|        |      Doc-Id | Version          | Created           | Modified           |                   |
|        | {{{DocID}}} | {{{DocVersion}}} | {{{CreatedDate}}} | {{{ModifiedDate}}} |                   |
[fn:1]

#+ATTR_ODT: :inner-frame (anchor "as-char")
#+ATTR_ODT: :width 2
[[./org-mode-unicorn.png]]

[fn:2]

#+begin_center
{{{DocTitle}}}
#+end_center


```

**Reply by hwiorn** [reply]
> I have added this feature. See #117 for more details.

Thank you for your fast patch! I'm checking out it now.
And how about general properties?
I think it has benefits for indexing like file search engine, spotlight to share documents to other people.

> If the document is in English, then the header will be too cramped. If you are using korean or other east-asian languages this may not be the case.

I'm making two version(english, korean) of the template now. I think it's not matter because the header font size is so small(8pt).

![image](https://user-images.githubusercontent.com/1394254/139568114-65fe01a0-af60-4f88-a395-4f4a591c4f4b.png)


> Note that I have used "transcluded table" (instead of list table).

I didn't know `transcluded table`. Is it a new feature of enhanced ox-odt? I can't find any reference of it.


And could you make a value type in custom fields? 

https://github.com/kjambunathan/org-mode-ox-odt/blob/e90550e5df70e6385eb3f70733ba4644cbd9948d/lisp/ox-odt.el#L2506-L2516

You hard-coded `string` type in custom field. I think it's better if there is specific value-type like `#+DOC-CREATED-DATE:     [2021-10-19 Tue] :type date`. Because ODT doesn't understand date type of org and always represents as string type.

**Reply by kjambunathan** [reply]
> I didn't know transcluded table. Is it a new feature of enhanced ox-odt? I can't find any reference of it.

Always search for what you are looking for in either the issues or code.   See [Add support for "Transcluded Tables" ](https://github.com/kjambunathan/org-mode-ox-odt/issues/112)


> And how about general properties?
> I think it has benefits for indexing like file search engine, spotlight to share documents to other people.

Could you tell me specific examples what you mean by general properties?  I need to see what fields are of interest to **you**, so that I can find out all that needs to be done to help you produce the document you are working on.

> You hard-coded string type in custom field. I think it's better if there is specific value-type like keyword-to-documentproperty-with-type(DOC-CREATED-DATE, date). Because ODT doesn't understand date type of org and always represents as string type.

I was anticipating this request ...  There is already good amount of code to deal with date fields, in timestamp ... including how the value may be displayed.


But I had to apply brakes because of a bug somewhere, and I couldn't get it to work.   (**If you could help me with isolating the issue, I will be happy. It will make things faster for you and easier for me.**)

These are good starting points

 See

https://github.com/kjambunathan/org-mode-ox-odt/blob/e90550e5df70e6385eb3f70733ba4644cbd9948d/lisp/ox-odt.el#L159

https://github.com/kjambunathan/org-mode-ox-odt/blob/e90550e5df70e6385eb3f70733ba4644cbd9948d/lisp/ox-odt.el#L2417

The bug is probably because of code-rot, i.e., the code might have diverged in the last decade or so and  I think it is because Emacs Date Time APIs have changed and ox.el or org.el haven't quite caught up with it.  There is no other exporter that handles date / datetime types like the ODT exporter does.   I think the issue is  somewhere here (or  the manner in which I or ox.el invoke it)

https://github.com/kjambunathan/org-mode-ox-odt/blob/e90550e5df70e6385eb3f70733ba4644cbd9948d/lisp/org.el#L20120


https://github.com/kjambunathan/org-mode-ox-odt/blob/e90550e5df70e6385eb3f70733ba4644cbd9948d/lisp/org.el#L20105

**As I requested, if you can debug a bit and help me with the root cause (or suggest a solution) I will appreciate it.  I am not comfortable reviewing others code, so a simple text description will suffice ... don't bother about a PR.  Whatever you choose to do, I will add support for value types in another week.**

**Reply by hwiorn** [reply]
> Always search for what you are looking for in either the issues or code.

I searched it in Issues and Google and I didn't find it. Maybe it was typo, I think.

> Could you tell me specific examples what you mean by general properties? I need to see what fields are of interest to you, so that I can find out all that needs to be done to help you produce the document you are working on.

General properties is below image I already posted it. Sorry, I don't know what the correct term is.

![image](https://user-images.githubusercontent.com/1394254/139083581-49bee0ba-8c06-4f41-b5dc-0d2911358d01.png)

They are already defined in [ODT specification](http://docs.oasis-open.org/office/v1.2/os/OpenDocument-v1.2-os-part1.html#__RefHeading__1415090_253892949)

- `Created`: `<meta:creation-date>2021-10-01T10:12:34.823961119</meta:creation-date>`
- `Modified`: `<dc:date>2021-11-01T10:12:34.823961119</dc:date>`
- `Last printed`: `<meta:print-date>2021-10-01T10:12:34.823961119</meta:print-date>`
- `Revision number`: `<meta:editing-cycles>1</<meta:editing-cycles>`
- `Generator`: `<meta:generator>Org 9.5-??-888aaa9 enhanced-ox-odt 803ae54 </meta:generator>`

> But I had to apply brakes because of a bug somewhere, and I couldn't get it to work. (If you could help me with isolating the issue, I will be happy. It will make things faster for you and easier for me.)

I will check it :)

**Comment by kjambunathan** [comment]
Regarding date / datetime type fields,  I wonder what format you want the date value to be displayed.  Do korean users use different values of date or write them differently.  This is what I mean.

![Screenshot from 2021-10-31 10-57-26](https://user-images.githubusercontent.com/19869864/139569785-cb4962c2-9b34-446f-9f1d-a5c68b0fb723.png)

**Reply by hwiorn** [reply]
> Regarding date / datetime type fields, I wonder what format you want the date value to be displayed. Do korean users use different values of date or write them differently. This is what I mean.

Korean use the part of Western date format but not the same.

I need below date formats. "/" can be "-" or ignored in the documents. YYYY-MM-DD or YYYYMMDD.

- YY/MM/DD
- YYYY/MM/DD
- MM/DD
- MM/DD (NN)
- YY년 M월 D일
- YYYY년 M월 D일
- YY"년도 제"Q"사분기"
- M월 D일

Korean use hanja(kanji) and I think the date format is similar but not the same in Japan and Chinese(e.g. Era) .
You can see the Korean format more in LO directly.

![image](https://user-images.githubusercontent.com/1394254/139621990-54c43793-42a6-4997-88f4-fd1de8d8e098.png)

**Comment by kjambunathan** [comment]
Styles can now be specified using `nxml-mode` blocks.  This means one can use the full power of Org Babel to generate complex documents.  

For a sample document with auto-generated page header and footer, see https://github.com/kjambunathan/org-mode-ox-odt/discussions/180#discussioncomment-2749922
